source/common/upstream/ring_hash_lb.cc

```diff
@@ -60,34 +66,111 @@ HostConstSharedPtr RingHashLoadBalancer::Ring::chooseHost(uint64_t h) const {
-  if (hosts.empty()) {
-    return;
+
+  // Sanity-check that the locality weights, if provided, line up with the hosts per locality.
+  if (locality_weights != nullptr) {
+    ASSERT(locality_weights->size() == hosts_per_locality.get().size());
   }
 
-  // Currently we specify the minimum size of the ring, and determine the replication factor
-  // based on the number of hosts. It's possible we might want to support more sophisticated
-  // configuration in the future.
   // NOTE: Currently we keep a ring for healthy hosts and unhealthy hosts, and this is done per
   //       thread. This is the simplest implementation, but it's expensive from a memory
   //       standpoint and duplicates the regeneration computation. In the future we might want
   //       to generate the rings centrally and then just RCU them out to each thread. This is
   //       sufficient for getting started.
+
   const uint64_t min_ring_size =
-      config ? PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.value(), minimum_ring_size, 1024) : 1024;
+      config
+          ? PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.value(), minimum_ring_size, DefaultMinRingSize)
+          : DefaultMinRingSize;
+  const uint64_t max_ring_size =
+      config
+          ? PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.value(), maximum_ring_size, DefaultMaxRingSize)
+          : DefaultMaxRingSize;
+  const absl::optional<uint64_t> configured_replication_factor =
+      config && config->has_replication_factor()
+          ? absl::optional<uint64_t>(config->replication_factor().value())
+          : absl::nullopt;
+
+  // Sanity-check ring size bounds.
+  if (min_ring_size > max_ring_size) {
+    throw EnvoyException(fmt::format("ring hash: minimum_ring_size ({}) > maximum_ring_size ({})",
+                                     min_ring_size, max_ring_size));
+  }
 
-  uint64_t hashes_per_host = 1;
-  if (hosts.size() < min_ring_size) {
-    hashes_per_host = min_ring_size / hosts.size();
-    if ((min_ring_size % hosts.size()) != 0) {
-      hashes_per_host++;
+  // Compute the "effective" weight of each host as the product of its own weight and the locality
+  // weight, if given. Sum these effective weights.
+  uint64_t weighted_sum = 0;
+  std::unordered_map<HostConstSharedPtr, uint32_t> effective_weights;
+  for (uint64_t i = 0; i < hosts_per_locality.get().size(); ++i) {
+    for (const auto& host : hosts_per_locality.get()[i]) {
+      auto host_weight = host->weight();
+      ASSERT(host_weight != 0);
+      // NOTE: Locality weight may be zero. In this case, all hosts in the locality are assigned
+      //       no load. When we build the ring below, such hosts will have no entries in the ring.
+      // TODO: When we move to C++17, change this to `locality_weights[i]` (i.e. use
+      //       std::shared_ptr::operator[]) rather than dereferencing locality_weights explicitly.
+      auto locality_weight = locality_weights == nullptr ? 1 : (*locality_weights)[i];
```

Comment:
So, in that case, why not drop the zero locality weighted host early, here, and not have them contribute to weighted sum etc?

Reply:
Yup, that's fair.